src/Graph/GraphPrinter.php

```diff
@@ -133,79 +146,102 @@ public function getDependencyError() {
 	 *
 	 */
-	protected function processResultRow( array /* of SMWResultArray */ $row ) {
+	protected function processResultRow( array $row ) {
+		$node = null;
+		$fields = [];
+		$parents = [];
 		// loop through all row fields
-		foreach ( $row as $i => $resultArray ) {
-
-			// loop through all values of a multivalue field
-			while ( ( /* SMWWikiPageValue */
-				$object = $resultArray->getNextDataValue() ) !== false ) {
-
-				$type = $object->getTypeID();
-				if ( in_array( $type, [ '_wpg', '_wpp', '_wps', '_wpu', '__sup', '__sin', '__suc', '__con' ] ) ) {
-					// This is a page. A new node and an edge have to be created.
-					// create SRF\GraphNode for column 0
-					if ( $i === 0 ) {
+		foreach ( $row as $result_array ) {
+			$request = $result_array->getPrintRequest();
+			$type = $request->getTypeID();
+			// Whether this printout should be shown as an edge.
+			$show_as_edge = !$this->options->showGraphFields() // no fields at all.
+				|| in_array( $type, self::PAGETYPES ) // property of the type 'Page'.
+				|| $request->isMode( PrintRequest::PRINT_CHAIN ); // property chain, treated like 'Page'.
+
+			// Loop through all values of a multivalue field.
+			while ( ( /* SMWWikiPageValue */ $object = $result_array->getNextDataValue() ) !== false ) {
+				if ( $show_as_edge ) {
+					// All properties are shown as edges anyway OR this is a property of the type 'Page'.
+					if ( !$node && !$object->getProperty() ) {
+						// The graph node for the current record has not been created,
+						// and this is the printout '?'. So, create it now.
 						$node = new GraphNode( $object->getShortWikiText() );
 						$node->setLabel( $object->getPreferredCaption() ?: $object->getText() );
-						$this->nodes[] = $node;
 					} else {
-						$node->addParentNode(
-							$resultArray->getPrintRequest()->getLabel(),
-							$object->getShortWikiText()
-						);
+						// Remember a parent node to add after the row is processed.
+						$parents[] = [
+							'predicate' => $request->getLabel(),
+							'object' => $object->getShortWikiText()
+						];
 					}
 				} else {
-					// A non-page property.
-					$node->addField(
-						$resultArray->getPrintRequest()->getOutputFormat(),
-						$object->getShortWikiText(),
-						$type,
-						$resultArray->getPrintRequest()->getLabel()
-					);
+					// A non-Page property and 'graphfields' is set,
+					// so display it as a field after the row has been processed.
+					$fields[] = [
+						'name' => $request->getOutputFormat(),
+						'value' => $object->getShortWikiText(),
+						'type' => $type,
+						'page' => $request->getLabel()
+					];
 				}
 			}
 		}
+		// Add the node, if any, its parent nodes and fields for non-Page properties to the current edge.
+		if ( $node ) {
+			foreach( $parents as $parent ) {
+				$node->addParentNode( $parent['predicate'], $parent['object'] );
+				// @TODO: add explicit nodes with hyperlinks to every parent node not added as '?', but only once.
```

Comment:
Is this a lot of work?

Reply:
Yes, and this is neither a critical error, nor relevant to this task.